contracts/axelarnet-gateway/src/contract/execute.rs

```diff
@@ -253,3 +254,56 @@ fn try_load_executable_msg(
         None => Ok(None),
     }
 }
+
+/// Query Nexus module in core to generate an unique cross chain id.
+fn unique_cross_chain_id(client: &nexus::Client, chain_name: ChainName) -> Result<CrossChainId> {
+    let nexus::query::TxHashAndNonceResponse { tx_hash, nonce } =
+        client.tx_hash_and_nonce().change_context(Error::Nexus)?;
+
+    CrossChainId::new(
+        chain_name,
+        HexTxHashAndEventIndex::new(
+            tx_hash,
+            u32::try_from(nonce).change_context(Error::NonceOverflow)?,
+        ),
+    )
+    .change_context(Error::InvalidCrossChainId)
+}
+
+/// Query Nexus module in core to decide should route message to core
+fn determine_routing_destination(
+    client: &nexus::Client,
+    name: &ChainName,
+) -> Result<RoutingDestination> {
+    let dest = match client
+        .is_chain_registered(name)
+        .change_context(Error::Nexus)?
+    {
+        true => RoutingDestination::Nexus,
+        false => RoutingDestination::Router,
+    };
+
+    Ok(dest)
+}
+
+/// Route message to the Nexus module
+fn route_to_nexus(
+    client: &nexus::Client,
+    nexus: Addr,
+    msg: Message,
+    token: Option<Coin>,
+) -> Result<Response<nexus::execute::Message>> {
+    let msg: nexus::execute::Message = (msg, token.clone()).into();
+
+    token
+        .map(|token| BankMsg::Send {
+            to_address: nexus.to_string(),
+            amount: vec![token],
+        })
+        .map(Into::into)
+        .into_iter()
+        .chain(iter::once(client.route_message(msg)))
+        .collect::<Vec<_>>()
+        .then(|msgs| Response::new().add_messages(msgs))
+        .then(Ok)
```

Comment:
Don't really find this very readable imo. Maybe something more imperative but simpler instead?

```suggestion
    let mut response_msgs: Vec<CosmosMsg<nexus::execute::Message>> = vec![];

    if let Some(token) = token.clone() {
        response_msgs.push(BankMsg::Send {
            to_address: nexus.to_string(),
            amount: vec![token],
        }.into())
    };

    response_msgs.push(client.route_message((msg, token).into()));
    Response::new().add_messages(response_msgs).then(Ok)
```

Reply:
I like the current implementation better and more elegant.